File: HackerRank/Tutorials/Cracking the Coding Interview/Sorting_Bubble_sort.cpp
#include <iostream>
#include <vector>
using namespace std;

int main() {
    int n;
    cin >> n;
    vector<int> a(n);
    for(int a_i = 0;a_i < n;a_i++) {
        cin >> a[a_i];
    }
    int total_swaps = 0;
    for(int i = 0; i < n; ++i) {
        int number_of_swaps = 0;
        for(int j = 0; j < n - 1; ++j) {
            if (a[j] > a[j+1]) {
                total_swaps++;
                number_of_swaps++;
                a[j] ^= a[j+1];
                a[j+1] ^= a[j];
                a[j] ^= a[j+1];
            }
        }
        if (number_of_swaps == 0) {
            break;
        }
    }
    cout << "Array is sorted in " << total_swaps << " swaps." << endl;
    cout << "First Element: " << a[0] << endl;
    cout << "Last Element: " << a[n-1] << endl;
  return 0;
}

File: HackerRank/Tutorials/Cracking the Coding Interview/Sorting_Bubble_sort.java
import java.util.Scanner;

public class Solution {

    public static void main(String[] args) {
        /* Enter your code here.
           Read input from STDIN.
           Print output to STDOUT.
           Your class should be named Solution. */
        Scanner scan = new Scanner(System.in);
        int size = scan.nextInt();
        int[] arr = new int[size];
        for(int i = 0; i < size; ++i) {
            arr[i] = scan.nextInt();
        }
        scan.close();
        int totalSwaps = 0, numberOfSwaps;
        for(int i = 0; i < size; ++i) {
            numberOfSwaps = 0;
            for(int j = 0; j < size-1; ++j) {
                if (arr[j] > arr[j+1]) {
                    totalSwaps++;
                    numberOfSwaps++;
                    arr[j] ^= arr[j+1];
                    arr[j+1] ^= arr[j];
                    arr[j] ^= arr[j+1];
                }
            }
            if (numberOfSwaps == 0) {
                break;
            }
        }
        System.out.println("Array is sorted in " + totalSwaps + " swaps.");
        System.out.println("First Element: " + arr[0]);
        System.out.println("Last Element: " + arr[size-1]);
    }
}
